File: src/main/java/teamcity/crowd/plugin/loginmodule/CrowdLoginModule.java
package teamcity.crowd.plugin.loginmodule;

import com.atlassian.crowd.model.user.User;
import com.google.common.base.Optional;
import teamcity.crowd.plugin.PluginCrowdClient;
import teamcity.crowd.plugin.utils.LoggerFactory;

import javax.security.auth.Subject;
import javax.security.auth.callback.Callback;
import javax.security.auth.callback.CallbackHandler;
import javax.security.auth.callback.NameCallback;
import javax.security.auth.callback.PasswordCallback;
import javax.security.auth.login.FailedLoginException;
import javax.security.auth.login.LoginException;
import javax.security.auth.spi.LoginModule;
import java.util.Map;

public class CrowdLoginModule implements LoginModule {

    private CallbackHandler myCallbackHandler;
    private NameCallback myNameCallback;
    private PasswordCallback myPasswordCallback;
    private Callback[] myCallbacks;
    private Subject mySubject;

    private PluginCrowdClient pluginCrowdClient;
    private LoggerFactory loggerFactory;
    private LoggedInUserService loggedInUserService;

    @Override
    public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState, Map<String, ?> options) {
        myCallbackHandler = callbackHandler;
        myNameCallback = new NameCallback("login:");
        myPasswordCallback = new PasswordCallback("password:", false);
        myCallbacks = new Callback[]{myNameCallback, myPasswordCallback};
        mySubject = subject;
        pluginCrowdClient = (PluginCrowdClient) options.get(PluginCrowdClient.class.getName());
        loggedInUserService = (LoggedInUserService) options.get(LoggedInUserService.class.getName());
        loggerFactory = (LoggerFactory) options.get(LoggerFactory.class.getName());
    }

    @Override
    public boolean login() throws LoginException {
        try {
            myCallbackHandler.handle(myCallbacks);
        } catch (Exception e) {
            throw new LoginException(e.toString());
        }
        final String username = myNameCallback.getName();
        final String password = new String(myPasswordCallback.getPassword());

        String message = String.format("Attempting to log in with user [%s]", username);
        loggerFactory.getServerLogger().debug(message);

        Optional<User> possiblyLoggedInUser = pluginCrowdClient.loginUserWithPassword(username, password);
        if (possiblyLoggedInUser.isPresent()) {
            mySubject.getPrincipals().add(loggedInUserService.updateMembership(possiblyLoggedInUser.get()));
            return true;
        }
        throw new FailedLoginException("Invalid username or password");
    }

    @Override
    public boolean commit() throws LoginException {
        return true;
    }

    @Override
    public boolean abort() throws LoginException {
        return true;
    }

    @Override
    public boolean logout() throws LoginException {
        return true;
    }
}

File: teamcity-crowd-plugin-server/src/main/kotlin/teamcity/crowd/plugin/CrowdLoginModule.kt
package teamcity.crowd.plugin

import com.intellij.openapi.diagnostic.Logger
import javax.security.auth.Subject
import javax.security.auth.callback.Callback
import javax.security.auth.callback.CallbackHandler
import javax.security.auth.callback.NameCallback
import javax.security.auth.callback.PasswordCallback
import javax.security.auth.login.FailedLoginException
import javax.security.auth.login.LoginException
import javax.security.auth.spi.LoginModule


class CrowdLoginModule : LoginModule {
    private lateinit var myCallbackHandler: CallbackHandler
    private lateinit var myNameCallback: NameCallback
    private lateinit var myPasswordCallback: PasswordCallback
    private lateinit var myCallbacks: Array<Callback>
    private lateinit var mySubject: Subject
    private lateinit var pluginCrowdClient: PluginCrowdClient
    private lateinit var loggedInUserService: LoggedInUserService
    private lateinit var logger: Logger

    override fun initialize(subject: Subject?, callbackHandler: CallbackHandler?, sharedState: MutableMap<String, *>?, options: MutableMap<String, *>?) {
        myCallbackHandler = callbackHandler!!
        myNameCallback = NameCallback("login:")
        myPasswordCallback = PasswordCallback("password:", false)
        myCallbacks = arrayOf(myNameCallback, myPasswordCallback)
        mySubject = subject!!

        pluginCrowdClient = options!![ModuleDescriptorConstants.CROWD_CLIENT_OPTION] as PluginCrowdClient
        loggedInUserService = options[ModuleDescriptorConstants.LOGGED_IN_USER_SERVICE_OPTION] as LoggedInUserService
        logger = options[ModuleDescriptorConstants.LOGGER_OPTION] as Logger
    }

    override fun login(): Boolean {
        try {
            myCallbackHandler.handle(myCallbacks)
        } catch (e: Exception) {
            throw LoginException(e.toString())
        }
        val username = myNameCallback.name
        val password = String(myPasswordCallback.password)

        logger.debug("Attempting to log in with user [$username]")

        val user = pluginCrowdClient.loginUserWithPassword(username, password)
        if (user != null) {
            mySubject.principals.add(loggedInUserService.updateMembership(user))
            return true
        }
        throw FailedLoginException("Invalid username or password")
    }

    override fun commit(): Boolean {
        return true
    }

    override fun logout(): Boolean {
        return true
    }

    override fun abort(): Boolean {
        return true
    }

}